include/ghex/glue/atlas/atlas_user_concepts.hpp

```diff
@@ -0,0 +1,467 @@
+                    m_size{size} {}
+
+                // member functions
+                /** @brief generate halos
+                 * @param domain local domain instance
+                 * @return vector of receive halos, one for each PE*/
+                auto operator()(const domain_type& domain) const {
+
+                    std::vector<halo> halos{};
+                    for (int rank = 0; rank < m_size; ++rank) {
+                        halos.push_back({rank, domain.levels()});
+                    }
+
+                    const int* partition_data = atlas::array::make_view<int, 1>(domain.partition()).data();
+                    const index_t* remote_index_data = atlas::array::make_view<index_t, 1>(domain.remote_index()).data();
+
+                    // if the index refers to another rank, or even to the same rank but as a halo point,
+                    // corresponding halo is updated
+                    for (auto domain_idx = 0; domain_idx < domain.size(); ++domain_idx) {
+                        if ((partition_data[domain_idx] != m_rank) || (remote_index_data[domain_idx] != domain_idx)) {
+                            halos[partition_data[domain_idx]].push_back(domain_idx, remote_index_data[domain_idx]);
+                        }
+                    }
+
+                    return halos;
+
+                }
+
+        };
+
+        /** @brief CPU data descriptor*/
+        template <typename T, typename DomainDescriptor>
+        class atlas_data_descriptor {
+
+            public:
+
+                using value_type = T;
+                using index_t = typename DomainDescriptor::index_t;
+                using domain_id_t = typename DomainDescriptor::domain_id_type;
+                using arch_type = gridtools::ghex::cpu;
+                using device_id_type = gridtools::ghex::arch_traits<arch_type>::device_id_type;
+                using Byte = unsigned char;
+
+            private:
+
+                const DomainDescriptor& m_domain;
+                atlas::array::ArrayView<T, 2> m_values;
+
+            public:
+
+                atlas_data_descriptor(const DomainDescriptor& domain,
+                                      const atlas::Field& field) :
+                    m_domain{domain},
+                    m_values{atlas::array::make_view<T, 2>(field)} {}
+
+                /** @brief data type size, mandatory*/
+                std::size_t data_type_size() const {
+                    return sizeof (T);
+                }
+
+                domain_id_t domain_id() const { return m_domain.domain_id(); }
+
+                device_id_type device_id() const { return 0; }
+
+                /** @brief single access set function, not mandatory but used by the corresponding multiple access operator*/
+                void set(const T& value, const index_t idx, const std::size_t level) {
+                    m_values(static_cast<std::size_t>(idx), level) = value; // WARN: why std::size_t cast is needed here?
```

Comment:
what about the warning?

Reply:
It is on the Atlas side, I would keep it just to remember